Add XSSFOptimiser - XSSF equivalent of HSSFOptimiser

cowwoc opened the pull request:
## Summary

Adds `XSSFOptimiser` class to provide optimization methods for XSSF workbooks, equivalent to the existing `HSSFOptimiser` for HSSF workbooks.

This addresses the common "style explosion" problem where programmatically created XSSF workbooks accumulate too many duplicate cell styles and fonts, causing:
- Slow file opening in Excel
- Excel hangs or crashes with large workbooks
- Increased file size
- The "Too many different cell formats" error

## Changes

**New files:**
- `poi-ooxml/src/main/java/org/apache/poi/xssf/usermodel/XSSFOptimiser.java`
- `poi-ooxml/src/test/java/org/apache/poi/xssf/usermodel/TestXSSFOptimiser.java`

**Methods added:**
- `optimiseCellStyles(XSSFWorkbook)` - Remaps cells using duplicate styles to canonical (first) occurrence
- `optimiseFonts(XSSFWorkbook)` - Updates style font references to canonical fonts

## Implementation Notes

Unlike HSSF, XSSF's `StylesTable` maintains a separate Java list from the underlying `CTStylesheet` XML. Removing entries from `CTStylesheet` can cause `XmlValueDisconnectedException` on save because the Java objects become disconnected from their XML backing.

The implementation uses a **safe remapping approach**:
- Identifies duplicate styles/fonts by comparing all properties (not indices)
- Remaps cells to use canonical (first occurrence) styles
- Does NOT remove entries from XML (avoids disconnection issues)
- Prevents further style explosion

## Test Coverage

Includes 6 comprehensive tests:
- `testOptimiseCellStylesRemapsDuplicates`
- `testOptimiseCellStylesHandlesUnusedStyles`
- `testOptimiseCellStylesPreservesDifferentStyles`
- `testOptimiseFontsRemapsDuplicates`
- `testWorkbookSaveableAfterOptimization`
- `testOptimizationPreservesFormulas`

## Usage Example

```java
XSSFWorkbook workbook = new XSSFWorkbook(new FileInputStream("bloated.xlsx"));
XSSFOptimiser.optimiseFonts(workbook);
XSSFOptimiser.optimiseCellStyles(workbook);
workbook.write(new FileOutputStream("optimized.xlsx"));
```

Closes #951

cowwoc committed:
Add XSSFOptimiser - XSSF equivalent of HSSFOptimiser

Adds optimization methods for XSSF workbooks to address style explosion
issues. HSSFOptimiser exists for HSSF but no equivalent existed for XSSF.

Methods added:
- optimiseCellStyles(XSSFWorkbook): Remaps cells using duplicate styles
  to canonical (first) occurrence
- optimiseFonts(XSSFWorkbook): Updates style font references to canonical
  fonts

Implementation uses a safe remapping approach rather than removing entries
from XML, which avoids XmlValueDisconnectedException issues caused by
StylesTable's internal list becoming disconnected from CTStylesheet.

Includes comprehensive test coverage in TestXSSFOptimiser.

Closes #951

pjfanning commented on poi-ooxml/src/main/java/org/apache/poi/xssf/usermodel/XSSFOptimiser.java:
```diff
@@ -0,0 +1,247 @@
+            XSSFColor xssfColor = (XSSFColor) color;
+            byte[] rgb = xssfColor.getRGB();
+            if (rgb != null) {
+                return String.format("#%02X%02X%02X", rgb[0] & 0xFF, rgb[1] & 0xFF, rgb[2] & 0xFF);
```
forbidden api call - you need to use the version of string.format that takes a Locale as an input.

Locale.Root is fine

cowwoc replied:
Done

pjfanning commented:
Could you add `@since 6.0.0` on the new main source class? That will be the next non-bugfix release.

cowwoc committed:
Address PR review comments

- Add "since 6.0.0" javadoc tag to XSSFOptimiser class
- Use Locale.ROOT in String.format() to comply with forbidden-apis check

cowwoc commented:
@pjfanning Done. Try now.

pjfanning reviewed (approved):
lgtm - merged

pjfanning commented:
Just for documentation purposes: POI users are much better off arranging their code so that they create as few cell style instances as possible and reuse the instances as much as feasible. That is, don't keep creating the same style over and over. Create it once and cache a CellStyle instance in memory and use that for each cell that needs that style.
